File: problem_014.lua
#!/usr/bin/env lua

print("Project Euler - Problem 14:")
print("Which starting number, under one million, produces the longest chain by following the rules below?\n")
print("if n is even, n/2\nif n is odd, 3n+1\n")

limit = 1000000
history = {}
max = {0, 0}

for i = 1, limit, 1 do
   n = i
   count = 0

   -- Main loop to process number
   while (n ~= 1) do
      if (n % 2 == 0) then
         n = n / 2
      else
         n = 3 * n + 1
      end
      count = count + 1

      -- Refer to history if we have number less than i
      if (n < i) then
         count = count + history[n]
         n = 1
      end
   end
   history[i] = count

   -- Keep track of max result
   if (count > max[1]) then
      max[1] = count
      max[2] = i
   end
end

print(max[2])

File: problem_014.js
#!/usr/bin/node

console.log("Project Euler - Problem 14:");
console.log("Which starting number, under one million, produces the longest chain by following the rules below?\n");
console.log("if n is even, n/2");
console.log("if n is odd, 3n+1\n");

const limit = 1000000;
var longest = [0, 0];
var history = [0, 0];

for (var i=2; i<limit; i++) {
    let n = i;
    let chain = [];

    while (history.length <= n) {
	chain.push(n);
	if (n % 2 == 0)
	    n /= 2;
	else
	    n = (3 * n) + 1;
    }

    let total_length = history[n] + chain.length + 1;
    history.push(total_length);
    if (total_length > longest[1]){
	longest[0] = i;
	longest[1] = total_length;
    }
}

console.log(longest[0] + " (" + longest[1] + ")");
